rskj-core/src/main/java/co/rsk/peg/pegininstructions/PeginInstructionsProvider.java

```diff
@@ -0,0 +1,77 @@
+package co.rsk.peg.pegininstructions;
+
+import co.rsk.bitcoinj.core.BtcTransaction;
+import co.rsk.bitcoinj.script.ScriptChunk;
+import co.rsk.bitcoinj.script.ScriptOpCodes;
+import co.rsk.peg.InvalidOpReturnOutputException;
+import co.rsk.peg.NoOpReturnException;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PeginInstructionsProvider {
+
+    private static final Logger logger = LoggerFactory.getLogger(PeginInstructionsProvider.class);
+
+    public PeginInstructions buildPeginInstructions(BtcTransaction btcTx) throws
+        PeginInstructionsException, NoOpReturnException, InvalidOpReturnOutputException {
+
+        PeginInstructionsBase peginInstructions;
+        byte[] opReturnOutput = extractOpReturnData(btcTx);
+
+        if (opReturnOutput == null || opReturnOutput.length == 0) {
+            String message = "Empty OP_RETURN data found";
+            logger.debug("[buildPeginInstructions] {}", message);
+            throw new PeginInstructionsException(message);
+        }
+
+        int protocolVersion = PeginInstructionsBase.extractProtocolVersion(opReturnOutput);
+
+        switch (protocolVersion) {
+            case 1:
+                PeginInstructionsVersion1 peginInstructionsVersion1 = new PeginInstructionsVersion1(btcTx.getParams());
+                peginInstructionsVersion1.parse(opReturnOutput);
+                peginInstructions = peginInstructionsVersion1;
+                break;
+            default:
+                logger.debug("[buildPeginInstructions] Invalid protocol version given");
+                throw new PeginInstructionsException("Invalid protocol version");
+        }
+
+        return peginInstructions;
+    }
+
+    protected static byte[] extractOpReturnData(BtcTransaction btcTx) throws NoOpReturnException, InvalidOpReturnOutputException {
+        byte[] data = new byte[]{};
+        int opReturnOccurrences = 0;
+
+        logger.info("[extractOpReturnData] Getting OP_RETURN data for btc tx: {}", btcTx.getHash());
+
+        for (int i=0; i<btcTx.getOutputs().size(); i++) {
+            List<ScriptChunk> chunksByOutput = btcTx.getOutput(i).getScriptPubKey().getChunks();
+            if (chunksByOutput.get(0).opcode == ScriptOpCodes.OP_RETURN) {
+                if (chunksByOutput.size() > 1) {
+                    data = btcTx.getOutput(i).getScriptPubKey().getChunks().get(1).data;
+                    opReturnOccurrences++;
+                } else {
+                    opReturnOccurrences++;
+                    data = null;
```

Comment:
thrown `InvalidOpReturnOutputException`

Reply:
It is throwing PeginInstructionsException